File: Java/MoveZeroes.java
package Java;

public class MoveZeroes{
    public void moveZeroes(int[] nums) {
        // move all the zeroes at the end
        // keep order of nonzeroes

        // case 1: [0] => [0]
        // case 2: [1,0] => [1,0]
        // case3: [0,1,3] => [1,3,0]

        // case 4: [0,0,1,2] => [1,2,0,0]
        // case 5: [0,1,0,2,0,3] => [1,2,3,0,0,0]

        // 2 pointers

        /*
        pt1 points to 1st element
        pt2 is one step ahead of pt1
        check if value at pt1 == 0
            yes => check if value at pt2 != 0
                yes => swap values at both pointers and move pointers   forward
                no => only move the second pointer
            no => just move both pointers forward

        time complexity => O(n)
        space complexity => O(1)
        */

        // use a while loop with the condition that the value of the pt2 i lss than the length of the array

        int pt1 = 0;
        int pt2 = 1;

        while(pt2 < nums.length){
            // check value at first pointer
            if (nums[pt1] == 0){
                if (nums[pt2] != 0){
                    int temp = nums[pt1];
                    nums[pt1] = nums[pt2];
                    nums[pt2] = temp;
                    pt1++;
                    pt2++;
                }
                else{
                    pt2++;
                }
            }
            else{
                pt1++;
                pt2++;
            }
        }
    }

}

File: Python/moveZeroes.py
def moveZeroes(self, nums):
        # Brute force
        # for value in nums:
        #     if value == 0:
        #         nums.remove(value)
        #         nums.append(value)
                
        # return nums

        # optimized solution
        # 2 pointers approach
        
        # if list start with a 0 value, swap it with the value next to it only if that value != 0
        # one pointer points to the front elements while another points to other elements in list
        # everytime the other pointer points to a value != 0:
        # set the value of the element after value at the front pointer to be equal to the value at the other pointer
        # change the value at the other pointer to be = 0
        # increment both pointers       
        # if value at other pointer == 0, jsut move pointer one step forward

        # left = 0
        # right1 = 1

        
        # if 0 not in nums:
        #     return nums

        # while nums[0] == 0 and len(nums) > 1:
        #     #right1 = 1
            
        #     if nums[right1] != 0 and right1 < len(nums):
        #         nums[0] = nums[right1]
        #         nums[right1] = 0
            
        #     if right1 == len(nums) - 1 and nums[right1] == 0:
        #         break
        #     right1 += 1

        # for right in range(1,len(nums)):
        #     if nums[right] != 0:
        #         nums[left+1] = nums[right]
        #         nums[right] = 0
        #         left += 1
        # return nums

        left = 0

        for right in range(len(nums)):
            if nums[right] != 0:
                temp = nums[left]
                nums[left] = nums[right]
                nums[right] = temp
                left += 1
        return nums